File: src/main/java/bad/robot/radiate/teamcity/SingleProjectMonitor.java
package bad.robot.radiate.teamcity;

import bad.robot.http.HttpClient;
import bad.robot.radiate.Aggregator;
import bad.robot.radiate.monitor.Information;
import bad.robot.radiate.monitor.MonitoringTask;
import bad.robot.radiate.monitor.NonRepeatingObservable;
import com.googlecode.totallylazy.Sequence;

import static bad.robot.radiate.Aggregator.aggregate;
import static com.googlecode.totallylazy.Sequences.sequence;
import static java.lang.String.format;
import static java.util.Collections.singletonList;

@Deprecated
public class SingleProjectMonitor extends NonRepeatingObservable implements MonitoringTask {

    private final HttpClient http;
    private final Server server;
    private final TeamCity teamcity;
    private final Project project;

    public SingleProjectMonitor(Project project, TeamCityConfiguration configuration) {
        this.project = project;
        this.server = new Server(configuration.host(), configuration.port());
        this.http = new HttpClientFactory().create(configuration);
        this.teamcity = new TeamCity(server, configuration.authorisation(), http, new JsonProjectsUnmarshaller(), new JsonProjectUnmarshaller(), new JsonBuildUnmarshaller());
    }

    @Override
    public void run() {
        try {
            Iterable<BuildType> buildTypes = teamcity.retrieveBuildTypes(singletonList(project));
            Sequence<Build> builds = sequence(buildTypes).mapConcurrently(teamcity::retrieveLatestBuild);
            Aggregator aggregated = aggregate(builds);
            notifyObservers(aggregated.activity(), aggregated.progress());
            notifyObservers(aggregated.status());
            notifyObservers(new Information(toString()));
        } catch (Exception e) {
            notifyObservers(e);
        }
    }

    @Override
    public String toString() {
        return format("%s (%s)", project.getName(), project.getId());
    }

}

File: src/main/scala/bad/robot/radiate/teamcity/SingleProjectMonitor.scala
package bad.robot.radiate.teamcity

import bad.robot.radiate.Aggregate.aggregate
import bad.robot.radiate.config.Config
import bad.robot.radiate.monitor.{Information, MonitoringTask, NonRepeatingObservable}
import bad.robot.radiate.teamcity.ListBuildTypesSyntax._

import scalaz.syntax.either._

class SingleProjectMonitor(project: Project, config: Config) extends NonRepeatingObservable with MonitoringTask {
  private val http = HttpClientFactory().create(config)
  private val teamcity = new TeamCity(new TeamCityUrl(config.teamcity.server.url), config.teamcity.server.authorisation, http, new JsonProjectsUnmarshaller, new JsonProjectUnmarshaller, new JsonBuildUnmarshaller)

  def run() {
    val builds = for {
      buildTypes  <- teamcity.retrieveBuildTypes(List(project))
      builds      <- buildTypes.getBuilds(teamcity)
      aggregation <- aggregate(builds).right
    } yield aggregation

    builds.fold(error => {
      notifyObservers(error)
    }, aggregation => {
      notifyObservers(aggregation.activity, aggregation.progress)
      notifyObservers(aggregation.status)
      notifyObservers(new Information(toString))
    })
  }

  override def toString = s"${project.name} (${project.id})"
}